Add WG Batch with charter (#6299)
kubernetes/community: wg-batch/charter.md

# WG Batch Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [sig-governance].

## Scope

Discuss and enhance the support for Batch (eg. HPC, AI/ML, data analytics) workloads
in core Kubernetes. We want to unify the way users deploy batch workloads to improve
portability and to simplify supportability for Kubernetes providers.

### In scope

- Additions to the batch API group, currently including Job and CronJob resources
  that benefit batch use cases such as HPC, AI/ML and data analytics.
- Primitives for job-level queueing, not limited to the k8s Job resource.
- Primitives to control and maximize utilization of resources in fixed-size clusters (on-prem) and elastic clusters (cloud).
- Runtime and scheduling support for specialized hardware (GPUs, NUMA, RDMA, etc.)

### Out of scope

> - Uses of the batch API resources as support for serving workloads (eg. backups, upgrades, migrations)

Reviewer: ```suggestion
- Uses of the batch APIs as support for serving workloads (eg. backups, upgrades, migrations)
```

Author: Done

> - Proposals that reimplement the functionality of core kubernetes components (job-controller, kube-scheduler, cluster-autoscaler).

Reviewer: I would say a replacement is in scope, which may result in reimplementing a controller, for example a v2 Job api.

Author: Ah, I was referring to forks that are supposed to run alongside.

Author: rephrased

> - Job workflows or pipelines. Third party frameworks serve these use cases without needing additional Kubernetes primitives.

Reviewer: +1, I think this is a well established area. But what is still in scope is potentially enhance job API to make it work better for external workflow frameworks, like Argo or kubeflow pipelines.

Author: rephrased

## Stakeholders

Stakeholders in this working group span multiple SIGs that own parts of the
code in core kubernetes components and addons.

- Apps
- Autoscaling
- Node
- Scheduling

## Deliverables

The list of deliverables is TBD, but here are some high level features:

- To SIG Apps:
  - Updated Job API that fulfills the needs of a wider range of batch applications.
  - A performant job controller that can scale to thousands of pods per minute.
- To SIG Scheduling and Autoscaling

>   - A set of APIs to support job queueing, a framework to support different queing policies and a ready-to-use implementation as a subproject.

Reviewer: also job-level provisioning

Author: added

  - Scheduling plugin(s) to support different batch needs.
- To SIG Node:
  - Runtime support for specialized hardware.

## Roles and Organization Management

> This wg follows adheres to the Roles and Organization Management outlined in [wg-governance]

Reviewer: ```suggestion
This WG adheres to the Roles and Organization Management outlined in [wg-governance]
```

Author: done

# WG Batch Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [sig-governance].

## Scope

Discuss and enhance the support for Batch (eg. HPC, AI/ML, data analytics) workloads
in core Kubernetes. We want to unify the way users deploy batch workloads to improve
portability and to simplify supportability for Kubernetes providers.

### In scope

- Additions to the batch API group, currently including Job and CronJob resources
  that benefit batch use cases such as HPC, AI/ML and data analytics.
- Primitives for job-level queueing, not limited to the k8s Job resource.
- Primitives to control and maximize utilization of resources in fixed-size clusters (on-prem) and elastic clusters (cloud).

> - Runtime and scheduling support for specialized hardware (GPUs, NUMA, RDMA, etc.)

Reviewer 2: `Multi-cluster scheduling for batch workloads.` 【For discussion】

From the feedback of my communication with some end users. The need for multi-cluster scheduling is important in large-scale reinforcement learning and simulation.The resources of a single cluster cannot meet the requirements of this type of batch job.

Author: I foresee multi-cluster support becoming an important aspect. However, with the purpose of keeping a narrow set of goals we can deliver, should we defer this for later? Although we could list it from the beginning so that we keep it in mind, but not necessarily prioritize it just yet. Thoughts?

Reviewer: I think it is useful to mention it to also grab the attention of sig multi-cluster

Reviewer 2: SG to me for `a narrow set of goals at beginning`

Author: I added a sentence related to multi-cluster.

Reviewer 3: whatever outcome exists in the single-cluster construct should be a two-way door for additive mutli-cluster solutions

### Out of scope

- Uses of the batch APIs as support for serving workloads (eg. backups,
  upgrades, migrations).
- Proposals that duplicate the functionality of core kubernetes components
  (job-controller, kube-scheduler, cluster-autoscaler).
- Job workflows or pipelines. Mature third party frameworks serve these
  use cases with the current kubernetes primitives. But additional primitives
  to support these frameworks could be in scope.

## Stakeholders

Stakeholders in this working group span multiple SIGs that own parts of the
code in core kubernetes components and addons.

- Apps
- Autoscaling
- Node
- Scheduling

## Deliverables

The list of deliverables is TBD, but here are some high level features:

- To SIG Apps:
  - Updated Job API that fulfills the needs of a wider range of batch applications.

>   - A performant job controller that can scale to thousands of pods per minute.

Reviewer 4: Cool! Serverless may benefit from this.

Reviewer 5: This will probably also need to result in node-level performance enhancements too.

Author: I don't think node-level performance improvements are a exclusive need for batch. I think they can be looked at by SIG Node independently. WDYT?

Reviewer: can this be resolved?

Reviewer 6: @ahg-g i think so. @endocrimes let's tackle those separately as needed.

- To SIG Scheduling and Autoscaling
  - A set of APIs to support job queueing, a framework to support different

>     queing policies and a ready-to-use implementation as a subproject.

Reviewer 2: ```suggestion
    queueing policies and a ready-to-use implementation as a subproject.
```

Reviewer 7: I'm guessing that the new APIs might well be CRDs? Is there any expectation that these become built-in APIs?

Reviewer: Yes, at least that is what we would be aiming for.

Author: Typo fixed.

# WG Batch Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [sig-governance].

> 

Author: Fixed

  use cases with the current kubernetes primitives. But additional primitives
  to support these frameworks could be in scope.

## Stakeholders

Stakeholders in this working group span multiple SIGs that own parts of the
code in core kubernetes components and addons.

- Apps
- Autoscaling
- Node
- Scheduling

## Deliverables

The list of deliverables is TBD, but here are some high level features:

- To SIG Apps:
  - Updated Job API that fulfills the needs of a wider range of batch applications.
  - A performant job controller that can scale to thousands of pods per minute.
- To SIG Scheduling and Autoscaling
  - A set of APIs to support job queueing, a framework to support different
    queueing policies and a ready-to-use implementation as a subproject.
  - Scheduling plugin(s) to support different batch needs.
- To SIG Autoscaling:
  - Capabilities for job-level provisioning.
- To SIG Node:
  - Runtime support for specialized hardware.

## Roles and Organization Management

This wg adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

> [wg-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/wg-governance.md

Author: Done

  use cases with the current kubernetes primitives. But additional primitives
  to support these frameworks could be in scope.

## Stakeholders

Stakeholders in this working group span multiple SIGs that own parts of the
code in core kubernetes components and addons.

- Apps
- Autoscaling
- Node
- Scheduling

## Deliverables

The list of deliverables is TBD, but here are some high level features:

- To SIG Apps:
  - Updated Job API that fulfills the needs of a wider range of batch applications.
  - A performant job controller that can scale to thousands of pods per minute.
- To SIG Scheduling and Autoscaling
  - A set of APIs to support job queueing, a framework to support different
    queueing policies and a ready-to-use implementation as a subproject.
  - Scheduling plugin(s) to support different batch needs.
- To SIG Autoscaling:
  - Capabilities for job-level provisioning.
- To SIG Node:
  - Runtime support for specialized hardware.

## Roles and Organization Management

This wg adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

> 

Reviewer 8: ```suggestion

To maintain a solid communication line between the Kubernetes groups and the wider CNCF community, KubeCon/CloudNativeCon maintainer track call for proposal submissions are required; if not selected, a video update would be required and listed on the working groups README
```

Reviewer 8: open to suggestion here from others :)

Author: Added as additional commitments

  - Primitives to control and maximize utilization of resources in fixed-size clusters (on-prem) and elastic clusters (cloud).
  - Runtime and scheduling support for specialized hardware (GPUs, NUMA, RDMA, etc.)
- To congregate leads and users from different external projects and user groups
  (CNCF TAGs, conferences and workshops) in the batch ecosystem to gather requirements,
  validate designs and encourage reutilization of core kubernetes APIs.

### Out of scope

- Enhancements with the following nature:
  - Addition of new API kinds that serve a specialized type of workload. The focus
    should be on general APIs that specialized controllers can build on top of.
  - Uses of the batch APIs as support for serving workloads (eg. backups,
    upgrades, migrations). These can be served by existing SIGs.
  - Proposals that duplicate the functionality of core kubernetes components
    (job-controller, kube-scheduler, cluster-autoscaler).
  - Job workflows or pipelines. Mature third party frameworks serve these
    use cases with the current kubernetes primitives. But additional primitives
    to support these frameworks could be in scope.

> - To dictate a _one-true-way_ to support batch workloads in Kubernetes.

Reviewer: k8s by design is extensible, so I am not sure how to evaluate what we propose against this criteria. Any new API we propose could be seen as "dictating" something.

@liggitt what is the suggestion from steering committee here to avoid exactly?

# WG Batch Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

Discuss and enhance the support for Batch (eg. HPC, AI/ML, data analytics, CI)
workloads in core Kubernetes. We want to unify the way users deploy batch
workloads to improve portability and to simplify supportability for Kubernetes
providers.

### In scope

> - The following enhancements:

Reviewer: done.

  - Additions to the batch API group, currently including Job and CronJob resources
    that benefit batch use cases such as HPC, AI/ML, data analytics and CI.
  - Primitives for job-level queueing, not limited to the k8s Job resource. Long-term,
    this could include multi-cluster support.
  - Primitives to control and maximize utilization of resources in fixed-size clusters 
    (on-prem) and elastic clusters (cloud).
  - Runtime and scheduling support for specialized hardware (GPUs, NUMA, RDMA, etc.)

> - To reduce fragmentation in the k8s batch ecosystem: congregate leads and users from 

Reviewer: done.

- Apps
- Autoscaling
- Node
- Scheduling

## Deliverables

The list of deliverables include the following high level features:

- To SIG Apps:
  - Updated Job API that fulfills the needs of a wider range of batch applications.
  - A performant job controller that can scale to thousands of pods per minute.
- To SIG Scheduling and Autoscaling
  - A set of APIs to support job queueing, a framework to support different
    queueing policies and a ready-to-use implementation as a subproject.
  - Scheduling plugin(s) to support different batch needs.
- To SIG Autoscaling:
  - Capabilities for job-level provisioning.
- To SIG Node:
  - Runtime support for specialized hardware.

## Roles and Organization Management

This wg adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

[wg-governance]: /committee-steering/governance/wg-governance.md

Additionally, the wg commits to:

- maintain a solid communication line between the Kubernetes groups and the wider CNCF community;
- submit a proposal to the KubeCon/CloudNativeCon maintainers track; if not selected, a video update will be recorded and listed below.

> ## Timelines and Disbanding

Reviewer: @dims @liggitt updated the exit criteria, please take a look.

> As a first mandate, the wg will define a roadmap in the first quarter

Reviewer 9: I think this (1) is most in the spirit of WGs generally, and I hope that we can work hard together to support this outcome.

Reviewer: I rephrased a bit to clarify that the focus is on early exit, but left the other options.

Reviewer 6: thanks @ahg-g

    that benefit batch use cases such as HPC, AI/ML, data analytics and CI.
  - Primitives for job-level queueing, not limited to the k8s Job resource. Long-term,
    this could include multi-cluster support.
  - Primitives to control and maximize utilization of resources in fixed-size clusters 
    (on-prem) and elastic clusters (cloud).
  - Runtime and scheduling support for specialized hardware (GPUs, NUMA, RDMA, etc.)
- To reduce fragmentation in the k8s batch ecosystem: congregate leads and users from 

>   different external and internal projects and user groups (CNCF TAGs, k8s working grouping

Author: working group?

I think the group is called Topology Aware Scheduling. However, it's not a WG under k8s. @swatisehgal what's the proper way of calling it?

Reviewer: changed to subproject

Reviewer 10: Yep, we have been referring to our group as Topology Aware Scheduling working group but it is not a formal WG under k8s so we can call it Topology Aware Scheduling subproject/team/task force. I think @ahg-g  already made updates to the text so we are good!

>   foucsed on batch-related features such as NUMA-aware scheduling) in the batch ecosystem to 

Author: typo: focused

Reviewer: done.
